src/modalities/optimizers/optimizer_factory.py

```diff
@@ -156,14 +156,45 @@ def _create_optimizer_groups(
             f"model {type(model)} has no parameters with requires_grad=True (i.e., no traininable parameters)."
         )
 
-    optimizer_groups = [
+    optimizer_groups = _built_optimizer_groups_via_weight_decay_split(
+        weight_decay, weight_decay_groups_excluded, weight_decay_groups, params
+    )
+    return optimizer_groups, ["with_weight_decay", "without_weight_decay"]
+
+
+def _built_optimizer_groups_via_weight_decay_split(
+    weight_decay: float,
+    weight_decay_groups_excluded: list[str],
+    weight_decay_groups: dict[str, list[str]],
+    params: dict[str, nn.Parameter],
+) -> OptimizerGroups:
+    params_per_weight_decay_groups: list[dict[str, object]] = [
         {
             "params": _filter_params_for_weight_decay_group(params, regex_expressions=weight_decay_groups[group]),
-            "weight_decay": weight_decay if group not in weight_decay_groups_excluded else 0.0,
+            "exclude": group not in weight_decay_groups_excluded,
         }
         for group in weight_decay_groups.keys()
     ]
-    return optimizer_groups, weight_decay_groups.keys()
+
+    optimizer_groups: OptimizerGroups = [
+        {
+            "params": sum((p["params"] for p in params_per_weight_decay_groups if not p["exclude"]), []),
```

Comment:
maybe add a commment that `sum` combines the parameters lists here. Was not aware that you could overload sum like this.